### PR TITLE
Add classpath_depth to allow controlling transitive dependency depth

### PR DESCRIPTION
**Pull request by brian-mcnamara**
The work here it to add a `classpath_depth` option to the bazel project view. This configuration allows the classpath of the project to be reduced, where 1 is only the dependencies required for the targets imported to compile. Setting it less then 0 disables the filtering and the existing functionality to be used.

This allows functionality similar to the IJ bazel plugin, which excludes transitives and provides a partial classpath. I would also like this concept be brought into IJ, but starting here

### CODE REVIEW

**Review by plaird — APPROVED** [review]
Cool feature. Doesn't Eclipse complain though if it can't find necessary classes when it launches tests?

**Comment by brian-mcnamara** [comment]
> Cool feature. Doesn't Eclipse complain though if it can't find necessary classes when it launches tests?

Yes, good call out. I need to think though this one, I know IJ invokes bazel to remove the classpath in the IDE from causing issues, may have to do something similar, or figure out a way to store the actual runtime classpath outside Eclipse's project model.

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/BuildFileAndVisibilityDrivenProvisioningStrategy.java [hunk 1]
```diff
@@ -248,29 +247,17 @@ public Map<BazelProject, Collection<ClasspathEntry>> computeClasspaths(Collectio
...
+                var targets = projectTargets.stream().map(BazelLabel::toString).collect(joining(" + "));
                 // get all accessible targets based on visibility
                 Set<BazelLabel> allVisibleTargets = workspace.getCommandExecutor()
                         .runQueryWithoutLock(
                             new BazelQueryForLabelsCommand(
                                     workspaceRoot,
                                     format(
-                                        "kind(java_library, visible(%s, //...))",
-                                        projectTargets.stream().map(BazelLabel::toString).collect(joining(" + "))),
+                                        "kind(java_library, visible(%s, //...)) except kind(java_library, rdeps(//..., %s))",
```
Comment (brian-mcnamara) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/BuildFileAndVisibilityDrivenProvisioningStrategy.java:258]: Optimization I found, that as far as I can tell, results in the same thing. But could use a second pair of eyes

Comment (guw) [reply bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/BuildFileAndVisibilityDrivenProvisioningStrategy.java:258]: Yeah. good find.

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/projectview/BazelProjectView.java [hunk 2]
```diff
@@ -51,7 +51,8 @@ public record BazelProjectView(
         boolean discoverAllExternalAndWorkspaceJars,
         GlobSetMatcher externalJarsDiscoveryFilter,
         boolean shardSync,
-        int targetShardSize) {
+        int targetShardSize,
+        int importDepth) {
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/projectview/BazelProjectView.java:55]: This optimization applies to only two target provisioning strategies. Thus, I think it should not be a separate thing in the project view but part of `target_provisioning_settings` map.
https://github.com/salesforce/bazel-eclipse/blob/main/docs/common/projectviews.md#target_provisioning_settings

Can you revert this and just use it from there?

Also, doc update should be part of this PR. The `docs/common/projectviews.md `is the proper place. It should also briefly mention the negative impacts, i.e. possible compile errors, lost ability to run unit tests, missing sources during debugging.

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/projectview/BazelProjectFileReader.java [hunk 3]
```diff
@@ -404,6 +406,10 @@ private void parseProjectFile(Path bazelProjectFile, BazelProjectViewBuilder bui
                         // ignore deprecated
                         break;
                     }
+                    case "import_depth": {
+                        builder.importDepth = parseSectionBodyAsInt(rawSection, builder.importDepth);
+                        break;
+                    }
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/projectview/BazelProjectFileReader.java:412]: See comment below for avoiding this change.

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/ProjectPerTargetProvisioningStrategy.java [hunk 4]
```diff
@@ -78,6 +81,27 @@ public Map<BazelProject, Collection<ClasspathEntry>> computeClasspaths(Collectio
...
+                                        targetLabels,
+                                        importDepth,
+                                        targetLabels,
+                                        importDepth),
+                                    true,
+                                    format("Querying for depdendencies for projects: %s", targetLabels)))
+                        .stream()
+                        .map(BazelLabel::new)
+                        .collect(toSet());
+            }
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/ProjectPerTargetProvisioningStrategy.java:103]: Can this be moved into the parent base class? It seems to be shared between both strategies.

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/JavaAspectsClasspathInfo.java [hunk 5]
```diff
@@ -363,6 +367,10 @@ public Collection<ClasspathEntry> compute() throws CoreException {
 
         // Collect jars referenced by runtime deps
         for (TargetKey targetKey : runtimeDeps) {
+            if ((availableDependencies.size() > 0)
+                    && !availableDependencies.contains(new BazelLabel(targetKey.getLabel().toString()))) {
+                continue;
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/JavaAspectsClasspathInfo.java:372]: * I am not particular happy with the name. It does not convey the intent/purpose clear. Perhaps `runtimeDependencyIncludes` is better?
* Also, consider moving this into a method for readability (eg., `shouldExcludeRuntimeDependency`).
* Can you add comment explaining the motivation, i.e. is it memory impact, performance, etc.?

**Review by guw — CHANGES_REQUESTED** [review]

**Comment by guw** [comment]
@brian-mcnamara I thinkn it would be good to capture the motivation for this change. Perhaps we should open an issue and add a few screenshots showing before and/or after timings?

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/BaseProvisioningStrategy.java [hunk 6]
```diff
@@ -123,14 +125,61 @@ public abstract class BaseProvisioningStrategy implements TargetProvisioningStra
...
+     * given status.
+     *
+     * @param project
+     *            the project to create the marker at (must not be <code>null</code>)
+     * @param status
+     *            the status to create the marker for (must not be <code>null</code>)
+     * @return the created marker (never <code>null</code>)
+     * @throws CoreException
+     */
+    static IMarker createClasspathContainerProblem(BazelProject project, IStatus status) throws CoreException {
```
Comment (brian-mcnamara) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/BaseProvisioningStrategy.java:139]: I moved the validation logic into JavaAspectClasspathInfo, so marked this package protected, static. Happy to move this out into a helper class, let me know

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/JavaAspectsClasspathInfo.java [hunk 7]
```diff
@@ -373,11 +404,33 @@ public Collection<ClasspathEntry> compute() throws CoreException {
...
             }
         }
 
-        return result.values();
+        return new ClasspathHolder(
+                result.values(),
+                hasUnloadedEntries ? unloadedEntries.values() : Collections.emptyList());
+    }
+
+    private boolean entryAvailable(ClasspathEntry entry) throws CoreException {
```
Comment (brian-mcnamara) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/JavaAspectsClasspathInfo.java:422]: Need to come up with a better name, but this use to be in the Strategy, but moved it here to remove the list iteration that happened multiple times.

**Comment by brian-mcnamara** [comment]
I renamed `import_depth` to `runtime_import_depth` due to JetBrains BSP plugin using `import_depth` flag as a helper to specify the depth of source projects to pull into the project view during sync.

**Comment by guw** [comment]
hmm ... should we use `classpath_depth`?

**Comment by brian-mcnamara** [comment]
> hmm ... should we use `classpath_depth`?

Not opposed to that, the only thing that may be confusing is setting the value to 0, which in the current implementation means exclude all runtime dependencies, but with `classpath_depth`, 0 to me would indicate even compile time classpath is excluded. I could just subtract 1 from the input to get around this and make it more intuitive.

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainer.java [hunk 8]
```diff
@@ -49,14 +50,16 @@
...
 
     private final IPath path;
     private final IClasspathEntry[] classpath;
+    private final IClasspathEntry[] unloadedClasspath;
 
-    public BazelClasspathContainer(IPath path, IClasspathEntry[] classpath) {
+    public BazelClasspathContainer(IPath path, IClasspathEntry[] classpath, IClasspathEntry[] unloadedClasspath) {
         this.path = path;
         this.classpath = classpath;
+        this.unloadedClasspath = unloadedClasspath;
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainer.java:62]:
```suggestion
        this.classpath = requireNotNull(classpath);
        this.unloadedClasspath = requireNotNull(unloadedClasspath);
```

Can you wrap this into a `requireNotNull` (needs static import from Objects)?

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainer.java [hunk 9]
```diff
@@ -78,4 +90,13 @@ public int getKind() {
     public IPath getPath() {
         return path;
     }
+
+    /**
+     * A array of unloaded classpath entries.
+     *
+     * @return
+     */
+    public IClasspathEntry[] getUnloadedEntries() {
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainer.java:99]: Can we rename to `getTransitiveEntries`? I think that better communicates the purpose.

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainerSaveHelper.java [hunk 10]
```diff
@@ -103,8 +102,13 @@ static final class LibraryEntryReplace implements Serializable {
         }
 
         IClasspathEntry getEntry() {
-            return JavaCore.newLibraryEntry(path, sourceAttachmentPath, sourceAttachmentRootPath, //
-                accessRules, extraAttributes, exported);
+            return JavaCore.newLibraryEntry(
+                path,
+                sourceAttachmentPath,
+                sourceAttachmentRootPath, //
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainerSaveHelper.java:108]:
```suggestion
                sourceAttachmentRootPath,
```

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainerSaveHelper.java [hunk 11]
```diff
@@ -150,12 +154,16 @@ static final class ProjectEntryReplace implements Serializable {
         }
 
         IClasspathEntry getEntry() {
-            return JavaCore.newProjectEntry(path, accessRules, //
-                combineAccessRules, extraAttributes, exported);
+            return JavaCore.newProjectEntry(
+                path,
+                accessRules, //
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/BazelClasspathContainerSaveHelper.java:159]:
```suggestion
                accessRules,
```

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/ClasspathHolder.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+import com.salesforce.bazel.eclipse.core.model.discovery.classpath.ClasspathEntry;
+
+/**
+ * Represents a targets classpath as two parts, loaded being the classpath entries that are loaded into the project
+ * model, and unloaded being classpaths that are part of the target, but are not to be loaded in the project model.
+ *
+ * Note: Unloaded contains null values to keep the order between loaded and unloaded
+ */
+public record ClasspathHolder(Collection<ClasspathEntry> loaded, Collection<ClasspathEntry> unloaded) {
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/classpath/ClasspathHolder.java:31]: Can we rename to `compile` and `transitive` dependencies? I think that better communicates the design/intent.

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/BaseProvisioningStrategy.java [hunk 13]
```diff
@@ -564,6 +615,57 @@ protected void analyzeProjectInfo(BazelProject project, JavaProjectInfo javaInfo
...
+        var classpathDepth = 0;
+        try {
+            classpathDepth = Integer.parseInt(classpathDepthStr);
+        } catch (NumberFormatException e) {
+            LOG.warn(
+                "Invalid integer for target provisioning setting {} (falling back to default 0): {}",
+                CLASSPATH_DEPTH,
+                e.getMessage(),
+                e);
+        }
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/BaseProvisioningStrategy.java:643]: Might be worth moving this into it's own method for readability?

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/JavaAspectsClasspathInfo.java [hunk 14]
```diff
@@ -627,4 +654,40 @@ protected ClasspathEntry resolveProject(TargetKey targetKey) throws CoreExceptio
...
 
+    /**
+     * Allows filtering runtime dependencies based on the availability of the dependency in runtimseDependencyIncludes
+     * when it includes elements
+     *
+     * @param targetKey
+     *            the dependency to check
+     * @return true if runtimeDependencyAvailable is empty or it includes the dependency
+     */
+    private boolean runtimeDependencyAvailable(TargetKey targetKey) {
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/JavaAspectsClasspathInfo.java:665]:
```suggestion
    private boolean includeRuntimeDependencyAsCompileDependency(TargetKey targetKey) {
```

File: bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/JavaAspectsClasspathInfo.java [hunk 15]
```diff
@@ -627,4 +654,40 @@ protected ClasspathEntry resolveProject(TargetKey targetKey) throws CoreExceptio
...
+     * @return false if the entry should not be included in the classpath
+     * @throws CoreException
+     *             if validation fails
+     */
+    private boolean validateEntry(ClasspathEntry entry) throws CoreException {
+        if ((entry.getEntryKind() == IClasspathEntry.CPE_LIBRARY) && !isRegularFile(entry.getPath().toPath())) {
+            BaseProvisioningStrategy.createClasspathContainerProblem(
+                bazelProject,
+                Status.error(
+                    format("Library '%s' is missing. Please consider running 'bazel fetch'", entry.getPath())));
```
Comment (guw) [inline bundles/com.salesforce.bazel.eclipse.core/src/com/salesforce/bazel/eclipse/core/model/discovery/JavaAspectsClasspathInfo.java:684]: Is this missing a `return false` here?

**Review by guw — APPROVED** [review]

**Pull request merged**